lib/src/command/upgrade.dart

```diff
@@ -140,9 +167,61 @@ Consider using the Dart 2.19 sdk to migrate to null safety.''');
       summaryOnly: onlySummary,
       analytics: analytics,
     );
+
     _showOfflineWarning();
   }
 
+  /// Updates [changes] to contain changes that updates constraints in [pubspec]
+  /// to have their lower bound match the version in [packages].
+  ///
+  /// If packages to update where given on the commandline, only those are
+  /// tightened. Otherwise all packages are tightened.
+  ///
+  /// If dependency has already been updated in [changes], the update will apply
+  /// on top of that change.
```

Comment:
Couldn't this just be a function taking a `List<PackageRange>` and returning a `List<PackageRange>`.

Then maybe `_updatePubspec(List<PackageRange> target)` and it has to figure out on its own which have changed and which haven't.

Ofcourse, I see that if we wanted to make those changes we should really make the refactoring to `_updatePubspec` in a separate PR (and land it first).